Action Client: Abstract Base Class  w/ Concrete Implementation

tko-dev opened the pull request:
- Abstract base class to enable other students and KTH folks to see what is required in regards to implementation
- Parameterized go to setpoint client and node for drone (launch file still needed)

Commit:
Initial exploration work and concrete implementation

Working on ABC framework for server

Adding type hinting and further expanding wrapping

WIP concrete implementation

fixing errors

all generic functions needing implementation

Expanding abstractions

Copying over previous code and adjusting

Working through parameters for server node

Expanding parameters and filling in needs

Debugging and formatting WIP

Auto-formatting imports

Working through debugging

Removing mistaken types etc

Docs and code updating

Working through error case handling

Commit:
BUG: transform is not correct

Needs fixing next time

Commit:
Reworking to use pose stamped

Commit:
Debugging this transform nightmare

tko-dev commented:
### Transform Skill Issues
Alright I have no idea why this transform is not working so I have dug as deep as I can. So here is the dump of my issues to resolve and what I have looked
#### Scene Setup
First and foremost wanted to evaluate whether something is broken in the scene. This all looks good to me and I don't see a problem :heavy_check_mark: 
![image](https://github.com/user-attachments/assets/32f76092-5725-4f4b-9bb3-76b7bcbad9f3)

#### Transform Deep Dive
All the transforms I checked in this red box all have [0,0,0] translations between them.
![image](https://github.com/user-attachments/assets/4a79a978-8e4c-480d-a1c6-3a667e5c356f)

Evaluating how this transform looks them from `utm_33_V` all the way to `Quadrotor/base_link_gt` I am seeing a consistent transform around
```At time 562.465356852
- Translation: [1297.000, 1156.000, 1.943]
- Rotation: in Quaternion [-0.000, 0.000, 0.383, 0.924]
- Rotation: in RPY (radian) [-0.000, 0.000, 0.785]
- Rotation: in RPY (degree) [-0.000, 0.000, 44.996]
- Matrix:
  0.707 -0.707 -0.000 1297.000
  0.707  0.707  0.000 1156.000
 -0.000 -0.000  1.000  1.943
  0.000  0.000  0.000  1.000```

**This to me looks like a relative transformation from the UTM point to the drone and not the point of the drone in UTM coordinates. Further looking, I took the point from that should be exactly 0,0 in `utm_33_V` which is the reference point and when I transform that from lat-long to UTM with geodesy I get the right answer using my code. If I then apply a transform from `utm_33_V` to `Quadrotor/base_link` I get an answer what still looks like an easting and a northing:
```[DEBUG] [1744224128.988884073] [setpoint_client]: Position after transform:
Frame: Quadrotor/base_link_gt
Position:geometry_msgs.msg.Point(x=5071786.605483685, y=4152096.8772220244, z=8.807470223568167)
Orientationgeometry_msgs.msg.Quaternion(x=4.4919873620112434e-08, y=-3.7352047566052704e-08, z=-0.3826456096381471, w=0.9238951982907192)```

I grepped and searched through the code. I basically xeroxed the lat/long to utm code and put it in this node to double check my conversions are correct which they are.  Basically out of ideas on why the tf transform is not handling the utm offset and conversion

tko-dev commented on behaviours/go_to_setpoint/go_to_setpoint/setpoint_server.py:
```diff
@@ -0,0 +1,231 @@
+        else:
+            return True
+
+    def convert_to_utm(self, point: GeoPoint) -> PoseStamped:
+        point: utm.UTMPoint = utm.fromMsg(point)
+        pose_stamp = PoseStamped()
+        pose_stamp.pose.position = point.toPoint()
+        zone, band = point.gridZone()
+        pose_stamp.header.frame_id = f"utm_{zone}_{band}"
+        self.logger.debug(
+            f"Point Position from lat long:{self._str_posestamp(pose_stamp)}"
+        )
+        return pose_stamp
```
My conversion from geopoint to UTM vs the current implementation in the service

https://github.com/smarc-project/smarc2/blob/humble/utilities/utm_latlon_converter/utm_latlon_converter/converter_service_node.py#L35-L47

tko-dev commented on behaviours/go_to_setpoint/go_to_setpoint/setpoint_server.py:
```diff
@@ -0,0 +1,231 @@
+                f"{self.robot_name}/{self._target_frame_param}{self._frame_suffix}"
+            )
+
+    def transform_goal(self, pose_stamped: PoseStamped) -> PoseStamped:
+        """Provides transformed point from pose_stamped.header.frame_id to self.target_frame.
+
+        Raises:
+            TransformException when transformation fails allowing for caller to handle exception
+
+        Returns:
+            pose: pose in specified frame
+        """
+        t = self._tf_buffer.lookup_transform(
+            target_frame =self.target_frame,
+            source_frame =pose_stamped.header.frame_id,
+            time=Time(seconds=0),
+            timeout=Duration(seconds=2),
+        )
+        # based on ReadMe in repository
+        return do_transform_pose_stamped(pose_stamped, t)
```
Transform look up and conversion implementation

Commit:
Type Shenanigans, commenting, and readme basics

Commenting and organzing some func calls

Adding some basic readme stuff

broke out some of the type shenanigans

KKalem commented:
`utm`->`map_gt` should not be a 0 transform. But I can confirm that it is. Will fix.

KKalem commented:
> `utm`->`map_gt` should not be a 0 transform. But I can confirm that it is. Will fix.

Fixed in Assets master.

Commit:
Reformatting files after convo with Ozer

Reorganizing files

renaming files

replacing names

renaming files for base

Updating resource file

Updating package names

Updating naming of directories

Renaming and beginning refactor

Moving types

Commit:
Documentation and bugfix work

Renaming action

fixing compile bug

fixing names

Fixing compile issues

Fixing bugs with action server

Starting to write docs

Doc writing

More docs writing

More docs

updating docs and adding interface

Creating cancel callback pipeline

Updating docs and logs

Cancel framework bugs worked out

Running linters

KKalem commented on messages/smarc_mission_msgs/CMakeLists.txt:
```diff
@@ -23,7 +23,7 @@ rosidl_generate_interfaces(${PROJECT_NAME}
   "msg/GotoWaypoint.msg"
   "msg/MissionControl.msg"
   "action/Empty.action"
-  "action/GotoWaypoint.action"
```
Maybe dont replace this just yet, sam is still using the old stuff 🙃

tko-dev replied:
Whoops yep that was an accident

KKalem commented on behaviours/go_to_geopoint/test/test_copyright.py:
we kinda never use these default tests, nicer to not have them at all, _in case_ ONE DAY we actually write meaningful tests 🙃

tko-dev replied:
Roger removed auto test files. Removed my test file as well as it was stale and not useful already

Commit:
PR feedback fixes

Removing tests per feedback

Undoing accidental deletion

Updating readmes

Removing unnecessary readme entry

Commit:
Fixing frame issue

Fixing minor type anno

Changing tol band and log level

tko-dev commented:
Squashed a bunch of my commits cleaning up git history. Fixed my final lingering bug when I did final tests. Tested merge locally. Cleaned up final debug items like log level and phat tolerance band on position.

KKalem commented on behaviours/go_to_geopoint/go_to_geopoint/geopoint_server.py:
```diff
@@ -0,0 +1,355 @@
+from tf2_geometry_msgs import do_transform_pose_stamped
+from tf2_ros import Buffer, TransformException, TransformListener
+
+KM_TO_METER = 100
```
1km is 1000m :)
